o1vm/src/pickles/proof.rs

```diff
@@ -10,11 +11,87 @@ pub struct WitnessColumns<G, S> {
     pub selector: S,
 }
 
-pub struct ProofInputs<G: KimchiCurve> {
-    pub evaluations: WitnessColumns<Vec<G::ScalarField>, Vec<G::ScalarField>>,
+pub struct ProofInputs<F: Field> {
+    pub evaluations: WitnessColumns<Vec<F>, Vec<F>>,
+}
+
+pub struct NotInversedProofInputs<F: Field> {
+    pub scratch: [Vec<ToInverseOrNot<F>>; crate::interpreters::mips::witness::SCRATCH_SIZE],
+    pub instruction_counter: Vec<F>,
+    pub error: Vec<F>,
+    pub selector: Vec<F>,
+}
+
+impl<F: ark_ff::Field> From<NotInversedProofInputs<F>> for ProofInputs<F> {
+    // This function collects the values marked as to inverse from
+    // the proof input, performs a batch inversion on them,
+    // and returns a usable proof input
+    fn from(not_inversed: NotInversedProofInputs<F>) -> ProofInputs<F> {
+        // Initialising the result
+        let mut res = ProofInputs::new(not_inversed.scratch[1].len());
+        res.evaluations.error = not_inversed.error;
+        res.evaluations.instruction_counter = not_inversed.instruction_counter;
+        res.evaluations.selector = not_inversed.selector;
+
+        // Collecting the values to inverse
+        let to_inverse_map: &mut HashMap<_, F> = &mut HashMap::new();
+        for (i, scratch_i) in not_inversed.scratch.iter().enumerate() {
+            for (j, ref x_j) in scratch_i.iter().enumerate() {
+                match x_j {
+                    ToInverseOrNot::NotToInverse(_) => (),
+                    ToInverseOrNot::ToInverse(x) => {
+                        // the result is none as this key has not been inserted yet
+                        let _none = HashMap::insert(to_inverse_map, (i, j), *x);
+                    }
+                }
+            }
+        }
+
+        // Perform the inversion
+        // FIXME: we go trough vec to get the array required by arkworks
+        // This is ugly
+        let mut to_inverse_vec: Vec<F> = Vec::with_capacity(to_inverse_map.len());
+        for v in to_inverse_map.values() {
+            to_inverse_vec.push(*v);
+        }
+        let to_inverse_slice = to_inverse_vec.as_mut_slice();
+
+        ark_ff::batch_inversion::<F>(to_inverse_slice);
```

Comment:
Oh, so here's the batch inversion... Then I am not sure about what `to_field()` is being used for.

Reply:
For the legacy, but I can change that and remove to field